### PR TITLE
Allow for constructor call based creation of instances of Key_

### PR DESCRIPTION
**Pull request by noseglasses**
Currently, the `Key_` union is lacking constructors that enable creation either via a keyCode/flags pair or raw data. In most places in the firmware, `Key_` instances are created via C-style struct initialization using initializer lists. It would be more convenient/complete to allowalso for class-style constructor initialization, e.g.
```cpp
Key_ a(128, 128); // keyCode/flags
Key_ b(12222); // raw
```
This PR adds the necessary constructors and keeps the default constructor to conserve initializer list based construction.

### CODE REVIEW

File: src/key_defs.h [hunk 1]
```diff
@@ -25,6 +25,16 @@ typedef union Key_ {
   };
   uint16_t raw;
 
+  Key_() = default;
```
Comment (gedankenexperimenter) [inline src/key_defs.h]: Is this necessary? I thought the default constructor would be created automatically. Is that a difference between `class` and `union`?

Comment (noseglasses) [reply src/key_defs.h]: Yes it is. If you don't implement any constructor, the compiler synthesizes a default constructor for you. Otherwise, as in the PR, the default constructor will be missing. Then you use the `... = default;` syntax to inform the compiler to generate the default constructor anyway. 

FWIW, this can be used with other stuff that is auto-generated, like assignment operators, too. It's sibling is `... = delete;` which is used to suppress auto-generated stuff.

**Comment by obra** [comment]
This change made me a little uncomfortable and it took me a couple days to figure out _why_. 

I think I've finally got it. It adds a second syntax for something fairly simple for no reason other than style. By having two ways to do the same thing, we end up with code that's more complicated for a new user to understand. Unless there's some other reason that we should be switching to this style of constructor, I feel like this is probably a no-go. (If there is, please say so!)

-jesse

**Comment by gedankenexperimenter** [comment]
I have a reason. I was trying something out a couple of days ago, and got stumped for a while by narrowing conversion errors that would have been avoided if the object initialization hadn't been in curly braces. It was very frustrating for about half an hour there, and I ended up avoiding the problem by creating an otherwise-unneccesary local variable. I'd rather have a real constructor and not have to use the braces.

**Comment by obra** [comment]
That's a reason. Given that, I'd rather see a wholesale conversion at some point when we're making a serious breaking change.

**Comment by gedankenexperimenter** [comment]
> That's a reason. Given that, I'd rather see a wholesale conversion at some point when we're making a serious breaking change.

I'm all in favor of that.

**Comment by noseglasses** [comment]
Brace style initializations like 
```cpp
(Key) { k.keyCode, k.flags | CTRL_HELD }
```
are a reminiscence to C and only there for compatibility reasons to continue support for C struct initialization. This type of initialization is hardly mentioned in any book about C++. Ordinary constructors, on the other hand, are familiar to any C++ programmer.

This PR increases readability and does not break anything.

**Comment by noseglasses** [comment]
One more thing. The legacy brace initialization does not come without problems and is by no means intuitive. 

Please look at the following tiny example.
```cpp
#include <cstdio>
#include <cstdint>

union U {
   struct {
      uint8_t a_;
      uint8_t b_;
   };
   uint16_t c_;
};

int main() {
   // The following line produces a compile error when used on any architecture 
   //    where int is not equal to int8_t.
   // main.cpp: In function ‘int main()’:
   // main.cpp:13:22: error: narrowing conversion of ‘1313’ from ‘int’ to ‘uint8_t {aka unsigned char}’ inside { } [-Wnarrowing]
   //     U u1 = (U) { 1313 };
   U u1 = (U) { 1313 };
   printf("u1.a_ = %hhu, u1.b_ = %hhu, u1.c_ = %hu\n", u1.a_, u1.b_, u1.c_);
   
   U u2 = (U) { 13, 42 };
   printf("u2.a_ = %hhu, u2.b_ = %hhu, u2.c_ = %hu\n", u2.a_, u2.b_, u2.c_);
}
```
This means, that when (accidentally) omitting the second entry in the initializer list, the compiler tries to assign the remaining value to `.a_` rather than `.c_`, contrary to what one might expect.

The problem becomes even worse when we only switch the order of the members of the union.
```cpp
#include <cstdio>
#include <cstdint>

union U {
   uint16_t c_;
   struct {
      uint8_t a_;
      uint8_t b_;
   };
};

int main() {
   U u1 = (U) { 1313 };
   printf("u1.a_ = %hhu, u1.b_ = %hhu, u1.c_ = %hu\n", u1.a_, u1.b_, u1.c_);
   
   // The following line yields 
   // main.cpp: In function ‘int main()’:
   // main.cpp:16:24: error: too many initializers for ‘U’
   //     U u2 = (U) { 13, 42 };
   U u2 = (U) { 13, 42 };
   printf("u2.a_ = %hhu, u2.b_ = %hhu, u2.c_ = %hu\n", u2.a_, u2.b_, u2.c_);
}
```
Now, we cannot pass two initializer values for `a_` and `b_` anymore, but can only pass `c_`.
This is because the only possible initialization list is always generated by exclusively considering the first member in the union.

We can conclude that using unions with anonymous brace initializers is purely evil. This is another reason why C++ introduced properly typed constructors. Unions should either be used with proper constructors or, if inevitable, with C-style named initialization, e.g. as
```cpp
U a = (U){ .a_ = 13, .b_ = 42 };
U b = (U){ .c_ = 1535 };
```
No programmer who learned C++ from scratch would write the latter. In fact only few of them would know that it is actually possible. I programmed C++ for more than a decade until I first stumbled upon this syntax, rummaging through old C code.

Please note, that you were lucky that what you did worked so far as you combined `uint8_t` and `uint16_t` in a union on a 8-bit machine. With other data types and other bitness you could face all sorts of strange behavior or if lucky, only compiler errors. Unions are to be handled with extra care as they are a tiny gap in an otherwise fortified language.

**Comment by noseglasses** [comment]
This is actually something that really hurts me as the missing constructors prevent proper Python wrapping of `Key`.

**Comment by noseglasses** [comment]
Some more info about Problems with classes/unions lacking constructors.

Initializing a nested aggregate is tricky (see http://en.cppreference.com/w/cpp/language/aggregate_initialization).
```cpp
union U {
   uint16_t c_;
   struct {
      uint8_t a_;
      uint8_t b_;
   };
};

 // This is illegal
U a = (U){ .a_ = 13, .b_ = 42 };

 // This might work with some compilers, but still illegal
U a = (U){ { .a_ = 13, .b_ = 42 } };

// There is no legal way to initialize a_ and b_ with aggregate initialization. 
// Only if the struct in U would be named, initialization would become possible.

// Another solution is to swap the struct and c_ (as it is the way in union Key).
union U2 {
   struct {
      uint8_t a_;
      uint8_t b_;
   };
   uint16_t c_;
};

// This is ok (implicit aggregate initialization of first member).
// Note the extra braces.
U2 a = (U2){ { .a_ = 13, .b_ = 42 } }; 

// Illegal, no aggregate but only list initialization possible with operator new for 
// unions (see http://en.cppreference.com/w/cpp/language/new).
U2 *aPtr = new U2{ { .a_ = 13, .b_ = 42 } };

// The latter (dynamic allocation) can be a problem with virtual builds where 
// dynamic allocation can be useful in some occasions.
```

**Comment by jwakely** [comment]
```
// This is ok (implicit aggregate initialization of first member).
// Note the extra braces.
U2 a = (U2){ { .a_ = 13, .b_ = 42 } }; 
```

Strictly speaking that's not valid in any version of C++, "designated-initializers" are a C feature that some C++ compilers support but are not in ISO C++ (they should be supported in C++20 though).

**Comment by noseglasses** [comment]
Good point. Thanks.

File: src/key_defs.h [hunk 2]
```diff
@@ -25,6 +25,16 @@ typedef union Key_ {
...
+
+  constexpr Key_(uint8_t __keyCode, uint8_t __flags)
+    : keyCode(__keyCode), flags(__flags) {
+  }
+
+  constexpr Key_(uint16_t __raw)
```
Comment (jwakely) [inline src/key_defs.h]: Should this constructor be `explicit`?

Comment (noseglasses) [reply src/key_defs.h]: You seem to have good C++ knowledge. That's very nice.

On the one hand, I agree with your suggestion. The explicit keyword protects against undesired automatic type conversions. On the other hand, I suppose that if someone passes a single argument, she or he intents to pass raw data. By adding the explicit keyword, we would force users to cast literal constants to appropriate type, am I right? If so, we would cause extra work without much gain.

The only other use case I can imagine is that someone intents to pass only a keycode and wants the flags member to be initialized with a default value. That could be made possible by making the raw-value constructor explicit and adding a default value for the two-parameter constructor's flag parameter. That, however, would mean that the type of the first (only) argument would decide on the constructor that is going to be called. I don't feel too comfortable with such a solution. To me it seems quite error prone.

Comment (jwakely) [reply src/key_defs.h]: > By adding the explicit keyword, we would force users to cast literal constants to appropriate type, am I right?

Well currently (with no constructors) there's no implicit conversion from an integer to a `Key` and it doesn't seem to cause any problems. So while you're right that you'd need to say `(Key)99` or `Key(99)` or `Key{99}` to explicitly convert an integer to a `Key`, that doesn't seem to be needed anywhere today.

By adding a non-`explicit` constructor you make it possible to do e.g.

````
void some_function(const Key&);
// ...
some_function(99);
````

If that's a desirable property, don't make the constructor explicit.

> That could be made possible by making the raw-value constructor explicit and adding a default value for the two-parameter constructor's flag parameter. That, however, would mean that the type of the first (only) argument would decide on the constructor that is going to be called.

Yes, and if you use `int` or some other type instead of exactly `uint16_t` or `uint8_t` it would be ambiguous and fail  to compile.

> I don't feel too comfortable with such a solution. To me it seems quite error prone.

I agree.

Comment (noseglasses) [reply src/key_defs.h]: If I got you right, you agree that if everybody expects the following
```cpp
Key(42) == Key(uint16_t(42)) == Key{ .raw = 42 } /* C++-20 ;-) */
```
we should be better off without an `explict` single argument constructor.

Comment (jwakely) [reply src/key_defs.h]: The first two cases are explicit conversions anyway, so work the same whether the single argument constructor is `explicit` or not.

Let's forget I said anything and leave it non-explicit :-)

Comment (noseglasses) [reply src/key_defs.h]: > Let's forget I said anything and leave it non-explicit :-)

No, please, I find it important to care about such things. C++ features like the `explicit` keyword can help to prevent strange errors and, thus, can safe the community a lot of time and frustration.

> The first two cases are explicit conversions anyway

Just for my understanding: AFAIK, the literal `42` [is by default](http://www.cplusplus.com/doc/tutorial/constants/) an `int`  and does therefore require an implicit type conversion to become `uint16_t`. So `Key(42)` is not supposed to compile if
```cpp
explicit constexpr Key_(uint16_t __raw) ...
```
Am I missing something?

Comment (jwakely) [reply src/key_defs.h]: > No, please, I find it important to care about such things.

OK sure, I just didn't want to be a pain if it wasn't welcome :-)

> does therefore require an implicit type conversion to become `uint16_t`

No, `explicit` doesn't do that.

tl;dr whether the `Key` constructor is `explicit` controls whether there are implicit conversions to `Key` but it can't stop implicit conversions from `int` to `uint16_t` (or any other conversions between the built-in integer types).

Longer version: 

You can never stop `int` implicitly converting to other integer types, that's an unavoidable feature of C and C++. Making the `Key` constructor `explicit` means that this doesn't work:

    uint16_t num = 1;
    Key k = num;  // error, requires implicit conversion from `uint16_t` to `Key`

Instead you must explicitly create a `Key` using one of these forms:

    Key k = Key(num);    // constructor notation
    Key k = (Key)num;    // cast notation
    Key k(num);      // direct-initialization
    Key k{num};      // direct-list-initialization i.e. using {} not ()

and you can't just use an integer where a `Key` is required:

    void some_function(const Key&);
    // ...
    uint16_t num = 42;
    some_function(num);   // error, requires implicit conversion from uint16_t to Key

But all these examples behave exactly the same if we use `int num` instead of `uint16_t num`, because an `int` can always implicitly convert to `uint16_t`.

So making the `Key` constructor `explicit` prevents integers implicitly converting to `Key` objects, but doesn't change _which_ integers are valid, i.e. `Key(42)` and `Key((uint16_t)42)` are identical. They both do direct-initialization of a `Key` from the value `(uint16_t)42` (the first one just implicitly converts `(int)42` to `(uint16_t)42`, and _then_ calls the `Key` constructor).

Comment (noseglasses) [reply src/key_defs.h]: > OK sure, I just didn't want to be a pain if it wasn't welcome :-)

Definitely not. Your above explanations and the rest of our discussion are __very__ helpful. Thanks.

Comment (gedankenexperimenter) [reply src/key_defs.h]: I was thinking about this one again. Now that I understand what's going on with the implicit type conversion allowed by the lack of the `explicit` specifier, I'm wondering: if it's desirable to allow those implicit conversions, is there a good reason not to allow implicit conversions the other way by defining the cast operator `operator uint16_t()` (or perhaps `operatory uint16_t&()`)?

If both implicit casts are allowed, that would remove the need for comparison operator overloads entirely, wouldn't it?

Comment (jwakely) [reply src/key_defs.h]: No no no. If you allow implicit conversions in both directions you basically lose any type safety at all, and reasoning about the code can get much harder.

As I said originally, "Generally single-argument constructors should be `explicit` to avoid unwanted implicit conversions" i.e. implicit conversions are generally unwanted, and the source of bugs and confusion. Allowing them in one direction after carefully deciding it makes sense is OK, but that doesn't mean even more implicit conversions would be even better.

Comment (gedankenexperimenter) [reply src/key_defs.h]: I was just about to answer my own question. You'd still get some of the type safety: passing a parameter of type `Foo` to a function `f(Bar b)` still gives a compile error, even if both `Foo` and `Bar` allow implicit conversions to and from `int`, but a comparison operator will happily convert them both and allow the comparison.

We _would_ still get the ability to have a very small definition of two or more different types that are thin integer wrappers, and be able to define polymorphic functions like `f(Apple x)` and `f(Orange x)`, and assignments of one type to the other will still give a compilation error,  but those comparison operators…

Comment (gedankenexperimenter) [reply src/key_defs.h]: I'm still not convinced that it makes sense to allow implicit conversions at all, for the record.

Comment (jwakely) [reply src/key_defs.h]: One of the main problems with implicit conversions to `uint16_t` would be that suddenly all kinds of un-type-safe expressions would compile:

````
Key n = Key_N;
Key lctrl = Key_LeftControl;
Key newfile = lctrl + n;  // use "Ctrl-N" to open a file
Key file1 = lctrl + 1;  // use "Ctrl-1" to switch to first file
````

If you're going to allow implicit conversion back to integers you might as well forget the `Key` abstraction completely and just use plain integers.

This topic is covered in detail in classic books like [C++ Coding Standards](http://www.gotw.ca/publications/c++cs.htm) and [More Effective C++](https://www.aristeia.com/books.html).

Some discussion in the C++ guidelines:
[Avoid conversion operators](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ro-conversion)    
[By default, declare single-argument constructors explicit](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)

Comment (gedankenexperimenter) [reply src/key_defs.h]: That example also applies (and is more relevant to) implicit conversion _from_ integers _to_ `Key` objects, though (also, less relevant, `lctrl + n` is also not what you want). That's an argument for disallowing both implicit conversions.

Comment (gedankenexperimenter) [reply src/key_defs.h]: To make this more relevant, I'm thinking this may apply better to possible ways that key addressing might change. If we had two classes, each storing a single byte, and one of them represents the address of a keyswitch (`KeyAddr`), and the other represents the address of an LED (`LedAddr`), we might reasonably want a two functions called `setColor()`, one of which is used to set the LED color on a particular key, and one of which is meant to set the color of a particular LED. Consider that the indices for these two things are different, and some keyboards have LEDs that are not associated with particular keys.

Using the simplest approach – typedefs – doesn't work, because `KeyAddr` and `LedAddr` would actually be the same type.

Allowing implicit conversions in both directions would greatly simplify the code used to implement the two types, while making sure that the correct `setColor()` function gets called. It would allow programmers to make errors in comparison and arithmetic operations, however. That's a far cry from "lose any type safety at all", and it would mean that someone reading through the code wouldn't encounter pages of operator overloads.

I'm just not totally convinced that either option is strictly better than the other. I'm very skeptical of the idea that allowing the conversion in one direction is the best option, though. It's much more confusing if it appears to work in some situations, but not others than if it either always works or never works.

Comment (jwakely) [reply src/key_defs.h]: It's not my code, and I don't really have a horse in this race. I'm happy to offer my opinionated advice, but I'm not going to fight over it :smile:

But I do think implicit conversions in both directions is a serous design flaw. There's a reason all the C++ experts who write books advise against implicit conversions, even though they appear tempting and superficially useful.

**Comment by algernon** [comment]
Took me a while, but I got around to rebasing this on top of master, #527 is the result. I'm closing this one.

Thank you for the PR, and apologies for taking so very long to finally get back to it.
